Content.Server/Doors/Systems/DoorSystem.Airlock.cs

```diff
@@ -0,0 +1,82 @@
+using Content.Server.DeviceLinking.Events;
+using Content.Server.Power.Components;
+using Content.Server.Wires;
+using Content.Shared.Doors.Components;
+using Content.Shared.Interaction;
+using Content.Shared.Power;
+using Content.Shared.Wires;
+using Robust.Shared.Player;
+
+namespace Content.Server.Doors.Systems;
+
+public sealed partial class DoorSystem
+{
+    [Dependency] private readonly WiresSystem _wiresSystem = default!;
+
+    private void InitializeAirlock()
+    {
+        SubscribeLocalEvent<AirlockComponent, ComponentInit>(OnAirlockInit);
+        SubscribeLocalEvent<AirlockComponent, SignalReceivedEvent>(OnSignalReceived);
+        SubscribeLocalEvent<AirlockComponent, PowerChangedEvent>(OnPowerChanged);
+        SubscribeLocalEvent<AirlockComponent, ActivateInWorldEvent>(OnActivate);
+    }
+
+    private void OnAirlockInit(Entity<AirlockComponent> airlock, ref ComponentInit args)
+    {
+        if (TryComp<ApcPowerReceiverComponent>(airlock, out var receiverComponent))
+            _appearance.SetData(airlock, DoorVisuals.Powered, receiverComponent.Powered);
+    }
+
+    private void OnSignalReceived(Entity<AirlockComponent> airlock, ref SignalReceivedEvent args)
+    {
+        if (args.Port != airlock.Comp.AutoClosePort || !airlock.Comp.AutoClose)
+            return;
+
+        airlock.Comp.AutoClose = false;
+        Dirty(airlock);
+    }
+
+    private void OnPowerChanged(Entity<AirlockComponent> airlock, ref PowerChangedEvent args)
+    {
+        airlock.Comp.Powered = args.Powered;
+        Dirty(airlock);
+
+        if (TryComp<AppearanceComponent>(airlock, out var appearanceComponent))
+            _appearance.SetData(airlock, DoorVisuals.Powered, args.Powered, appearanceComponent);
+
+        if (!TryComp(airlock, out DoorComponent? door))
+            return;
+
+        if (args.Powered)
+            UpdateAutoClose(airlock, door: door);
+        else if (door.State == DoorState.Open)
+            DoorSystem.SetNextStateChange((airlock, door), null);
```

Comment:
```suggestion
        else if (door.State == DoorState.Open)
            // Don't auto close if the door became unpowered
            DoorSystem.SetNextStateChange((airlock, door), null);
```
Lost a valuable comment.